src/main/java/gregicadditions/machines/TileEntityProcessingArray.java

```diff
@@ -356,31 +370,70 @@ protected void multiplyInputsAndOutputs(List<CountableIngredient> newRecipeInput
+		protected RecipeMap findRecipeMapAndCheckValid(ItemStack machineStack) {
 
-			for(int slot = 0; slot < inputs.getSlots(); slot++) {
+			if(machineStack == null || machineStack.isEmpty()) {
+				return null;
+			}
 
-				// find tileentity
-				ItemStack wholeItemStack = inputs.getStackInSlot(slot);
-				String unlocalizedName = wholeItemStack.getItem().getUnlocalizedNameInefficiently(wholeItemStack);
+			String unlocalizedName = machineStack.getItem().getUnlocalizedNameInefficiently(machineStack);
+			String recipeMapName = findRecipeMapName(unlocalizedName);
 
-				if(unlocalizedName.contains("gregtech.machine") || unlocalizedName.contains("gtadditions.machine")) {
-					this.numberOfMachines = Math.min(16, wholeItemStack.getCount());
-					String trimmedName;
-					String voltage = unlocalizedName.substring(unlocalizedName.lastIndexOf(".") + 1);
-					trimmedName = unlocalizedName.substring(0, unlocalizedName.lastIndexOf("."));
 
-					//Checks if the tile entity is actually a machine
-					if(!GAEnums.voltageMap.containsKey(voltage))
-						continue;
+			//Check the machine against the Config blacklist
+			if(!findMachineInBlacklist(recipeMapName)) {
 
-					this.machineName = trimmedName.substring(trimmedName.lastIndexOf(".") + 1);
-					this.machineTierVoltage = GAEnums.voltageMap.get(voltage);
-					this.machineItemStack = wholeItemStack;
-					break;
+				RecipeMap<?> rmap = RecipeMap.getByName(recipeMapName);
+
+				//Find the RecipeMap of the MTE and ensure that the Processing Array only works on SimpleRecipeBuilders
+				//For some reason GTCE has specialized recipe maps for some machines, when it does not need them
+				if (rmap != null && (rmap.recipeBuilder() instanceof SimpleRecipeBuilder ||
+						rmap.recipeBuilder() instanceof IntCircuitRecipeBuilder ||
+						rmap.recipeBuilder() instanceof ArcFurnaceRecipeBuilder ||
+						rmap.recipeBuilder() instanceof CutterRecipeBuilder ||
+						rmap.recipeBuilder() instanceof UniversalDistillationRecipeBuilder)) {
+
+					return rmap;
 				}
 			}
-			return machineName;
+			return null;
+		}
+
+		public String findRecipeMapName(String unlocalizedName) {
+
+			String trimmedName = unlocalizedName.substring(0, unlocalizedName.lastIndexOf("."));
+			trimmedName = trimmedName.substring(trimmedName.lastIndexOf(".") + 1);
+
+			//For some reason, the Cutting saw's machine name does not match the recipe map's unlocalized name, so correct it
+			//Same with the Electric Furnace
+			if(trimmedName.equals("cutter")) {
+				trimmedName = "cutting_saw";
+			}
+			else if(trimmedName.equals("electric_furnace")) {
+				trimmedName = "furnace";
+			}
+			else if(trimmedName.equals("ore_washer")) {
+				trimmedName = "orewasher";
+			}
+
+			return trimmedName;
+		}
+
+		public boolean findMachineInBlacklist(String unlocalizedName) {
+
+			String[] blacklist = GAConfig.processingArray.machineBlackList;
+
+			return Arrays.asList(blacklist).contains(unlocalizedName);
```

Comment:
I think our limitation here stems from how we get the values from the configs API. If we know the array elements are sorted, we could leverage `Arrays.binarySearch` rather than creating a wrapping data structure just to look up where an element is, or even just manually do a linear search over the elements (Java unfortunately has no `indexOf` type method on array primitives). That said it's unlikely that linear search will have a discernible impact on overall performance unless a massive number of machines are put into the list.

Reply:
while we *are* stuck with an array, we can cache a copy of it as a hash set, which only has to be constructed once (or, at any rate, only each time the config is changed). however, it's probably true that the performance impact is negligible

Reply:
There's plenty of ways we could go about optimizing any given thing, but before budgeting our finite time and effort on it we should determine that it actually manifests as a meaningful performance bottleneck in practice.